Improvements of the build system

ptagl opened the pull request:
- Added a new `configure` argument to enable the `-Wall` flag during the compilation
- Added the `--use-clang` argument to `build.sh` to switch the compilation from `gcc` to `clang`
- Added some scripts to perform static code analysis with `cppcheck` and `oclint`

Commit:
Fixed the Zendoo compilation with -Werror flag

Commit:
Added scripts for static code analysis with cppcheck and oclint

Commit:
Added the --use-clang flag to conditionally build with CLANG

Commit:
Fixed Windows cross-compilation

Additionally restored CLIENT_VERSION_REVISION and CLIENT_VERSION_BUILD that have been accidentally touched by mistake.

drgora committed:
Allow easy cross-compilation with the CT_PREFIX env var

CT_PREFIX should be the prefix to cross-compilation tools, including everything up to the last part of the binary name.
E.g. if you have /path/to/x86_64-unknown-linux-gnu-gcc, use CT_PREFIX=/path/to/x86_64-unknown-linux-gnu-

drgora committed:
If compiling with clang, use it also to compile rust dependencies

drgora committed:
Update zeromq to version 4.3.4.

This fixes compilation on clang >= 10.

Commit:
Added documentation and improved linting scripts

drgora commented on doc/static-code-analysis.md:
```diff
@@ -0,0 +1,51 @@
+
+Clang
+-----------------
+By default Zen is compiled with GCC, but it's easy to switch to Clang by calling `zcutil/build.sh` with the flag `--use-clang`. Using Clang together with `wall` or `werror` tipically generates a different set of warnings compared with GCC.
```
I'd suggest to make it clear that this is true only for Linux.

ptagl replied:
Fixed with commit [3d4a66d4b49dcda0db45db5837374851f2f49505](https://github.com/HorizenOfficial/zen/pull/431/commits/3d4a66d4b49dcda0db45db5837374851f2f49505).

a-petrini commented on zcutil/build.sh:
```diff
@@ -138,11 +138,18 @@ then
 
+# If --use-clang is true, use Clang as the compiler (instead of gcc):
+CLANG_ARG='false'
+if [ "x${1:-}" = 'x--use-clang' ]; then
```
As we introduced the new `--use-clang` option to build.sh, I suggest to add a short description to the message returned by the `--help` option on the same file.

ptagl replied:
Fixed with commit [5fdf9320cc619cf830579efd5f4132189ed7d5da](https://github.com/HorizenOfficial/zen/pull/431/commits/5fdf9320cc619cf830579efd5f4132189ed7d5da).

a-petrini commented on qa/lint/extended-lint-cppcheck.sh:
```diff
@@ -0,0 +1,31 @@
+    exit 0
+fi
+
+function join_array {
```
This function was originally used in tandem with grep to filter some specific warnings. As we now rely on an explicit suppression list, this function is unused and should be removed.

ptagl replied:
Fixed with commit [01ebdbfb657ffdaa6f154b44a8add9a1b479a452](https://github.com/HorizenOfficial/zen/pull/431/commits/01ebdbfb657ffdaa6f154b44a8add9a1b479a452).

a-petrini commented on configure.ac:
```diff
@@ -208,6 +215,14 @@ if test "x$enable_debug" = xyes; then
 ERROR_CXXFLAGS=
+
+if test "x$enable_wall" = "xyes"; then
+  if test "x$CXXFLAG_WERROR" = "x"; then
```
We are checking if the compiler supports the `-wall` flag in case we request compilation with all warnings enabled, however we are checking against the wrong var `$CXXFLAG_WERROR`.

ptagl replied:
Fixed with commit [81359e450cd86d475f4cbe162afeec6570b392fd](https://github.com/HorizenOfficial/zen/pull/431/commits/81359e450cd86d475f4cbe162afeec6570b392fd).

Commit:
Added a help string for --use-clang in build.sh

Also removed the help strings for some old flags that are not used anymore.

Commit:
Added a note to the static analysis doc for "--use-clang" compatibility

Commit:
Fixed a check for the "-wall" flag in the configure script

Commit:
Remove inclusions and exclusions through grep from the linting scripts

We switched to using the configuration parameters supported by cppcheck and OcLint to filter the output.
